packages/core/core/src/services/document-service/events.ts

```diff
@@ -0,0 +1,73 @@
+import { curry } from 'lodash/fp';
+
+import { UID, Schema, Utils, Modules } from '@strapi/types';
+import { sanitize } from '@strapi/utils';
+
+import { getDeepPopulate } from './utils/populate';
+
+const ALLOWED_EVENTS = {
+  ENTRY_CREATE: 'entry.create',
+  ENTRY_UPDATE: 'entry.update',
+  ENTRY_DELETE: 'entry.delete',
+  ENTRY_PUBLISH: 'entry.publish',
+  ENTRY_UNPUBLISH: 'entry.unpublish',
+  ENTRY_DRAFT_DISCARD: 'entry.draft-discard',
+};
+
+type EventName = Utils.Object.Values<typeof ALLOWED_EVENTS>;
+
+const sanitizeEntry = async (
+  model: Schema.ContentType<any> | Schema.Component<any>,
+  entry: Modules.Documents.AnyDocument
+) => {
+  return sanitize.sanitizers.defaultSanitizeOutput(
+    {
+      schema: model,
+      getModel(uid) {
+        return strapi.getModel(uid as UID.Schema);
+      },
+    },
+    entry
+  );
+};
+
+/**
+ * Triggers an event.
+ *
+ * It will populate the entry if it is not a delete event.
+ * So the event payload will contain the full entry.
+ */
+const emitEntryEvent = async (
+  uid: UID.Schema,
+  eventName: EventName,
+  entry: Modules.Documents.AnyDocument
+) => {
+  const populate = getDeepPopulate(uid, {});
+  const model = strapi.getModel(uid);
+
+  const emitEvent = async () => {
+    // There is no need to populate the entry if it has been deleted
+    let populatedEntry = entry;
+    if (eventName !== 'entry.delete' && eventName !== 'entry.unpublish') {
+      populatedEntry = await strapi.db.query(uid).findOne({ where: { id: entry.id }, populate });
+    }
+
+    const sanitizedEntry = await sanitizeEntry(model, populatedEntry);
+
+    await strapi.eventHub.emit(eventName, {
+      model: model.modelName,
+      uid: model.uid,
+      entry: sanitizedEntry,
+    });
+  };
+
+  /**
+   * strapi.db.query might reuse the transaction used in the doc service request,
+   * so this is executed after that transaction is committed.
+   */
+  strapi.db.transaction(async ({ onCommit }) => onCommit(emitEvent));
```

Comment:
How confident are we with this ?

Reply:
on this working, or on the transaction being finished when this is executed?

I have seen this error being triggered on this code without the use of this.

Reply:
I think we can be confident on the execution part, I wonder about the onCommit on it looks legit :D